.github/workflows/semver-check.yaml

```diff
@@ -0,0 +1,131 @@
+            command: "cargo semver-checks"
+          - crate: "protocols/v2/subprotocols/mining"
+            command: "cargo semver-checks"
+          - crate: "protocols/v2/subprotocols/template-distribution"
+            command: "cargo semver-checks"
+          - crate: "protocols/v2/sv2-ffi"
+            command: "cargo semver-checks"
+          - crate: "protocols/v2/roles-logic-sv2"
+            command: "cargo semver-checks"
+          - crate: "protocols/v1"
+            command: "cargo semver-checks"
+          - crate: "utils/bip32-key-derivation"
+            command: "cargo semver-checks"
+          - crate: "utils/error-handling"
+            command: "cargo semver-checks"
+          - crate: "utils/key-utils"
+            command: "cargo semver-checks"
+          - crate: "roles/roles-utils/network-helpers"
+            command: "cargo semver-checks"
+          - crate: "roles/roles-utils/rpc"
+            command: "cargo semver-checks"
+          - crate: "roles/jd-client"
+            command: "cargo semver-checks"
+          - crate: "roles/jd-server"
+            command: "cargo semver-checks"
+          - crate: "roles/mining-proxy"
+            command: "cargo semver-checks"
+          - crate: "roles/pool"
+            command: "cargo semver-checks"
+          - crate: "roles/translator"
+            command: "cargo semver-checks"
+          - crate: "common"
+            command: "cargo semver-checks"
+
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v2
+
+      - name: Install Rust
+        uses: actions-rs/toolchain@v1
+        with:
+          toolchain: stable
+          override: true
+
+      - name: Cache Cargo registry
+        uses: actions/cache@v2
+        with:
+          path: ~/.cargo/registry
+          key: ${{ runner.os }}-cargo-registry-${{ hashFiles('**/Cargo.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-cargo-registry-
+
+      - name: Cache Cargo index
+        uses: actions/cache@v2
+        with:
+          path: ~/.cargo/git
+          key: ${{ runner.os }}-cargo-index-${{ hashFiles('**/Cargo.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-cargo-index-
+
+      - name: Install dependencies
+        run: sudo apt-get update && sudo apt-get install -y cmake
+
+      - name: Install cargo-semver-checks
+        run: cargo install cargo-semver-checks --locked
+
+      - name: Fetch all branches
+        run: git fetch --all
+
+      - name: Identify target branch
+        id: target_branch
+        run: echo "::set-output name=branch::$(git rev-parse --abbrev-ref --symbolic-full-name @{u} | cut -d'/' -f2)"
+
+      - name: Identify changed crates
+        id: changes
+        run: |
+          TARGET_BRANCH=$(echo "${{ steps.target_branch.outputs.branch }}")
+          echo "::set-output name=changed_crates::$(git diff --name-only origin/$TARGET_BRANCH...HEAD | grep '^crates/' | cut -d'/' -f2 | sort | uniq)"
```

Comment:
I'm doing some tests on my fork: https://github.com/plebhash/stratum/actions/runs/9962822250

it seems the entire matrix is running, regardless of this

![image](https://github.com/user-attachments/assets/f3383944-dbe7-478c-9712-32693de0c411)

Reply:
also:
![image](https://github.com/user-attachments/assets/b9302bda-a272-4fd3-b066-17f79a2a73c5)

Reply:
and with this, I come back to my original point:

we should avoid footguns by trying to overengineer and overoptimize things prematurely

by attempting to make things run faster, we ended up with something that was actually behaving in a way that completely diverged from the intented goal

I'm simplifying this workflow on my fork, I'll share it here once it's ready to be cherry-picked into this PR

Reply:
@Shourya742 here's a simplified approach https://github.com/plebhash/stratum/commit/11717da74006ad518a900e6bf86f044e24f1ffa7

here's a test run: https://github.com/plebhash/stratum/actions/runs/9963466632/job/27529566883

execution wasn't lightning fast and it didn't finish in 5s, but it gets the job done in a straightforward way

it behaves perfectly as expected: `framing_sv2` broke the CI because we recently updated its APIs and we still haven't bumped its version

my suggestion for this PR is to cherry-pick this commit... if we eventually realize this CI is prohibitively slow, we can investigate strategies for optimization

Reply:
Okay, I will let go of the filtering.